### PR TITLE
Ep

### PR DESCRIPTION
**Pull request by S1ro1**
Qwen3-30B-A3B w muon on wordle
image

> [!NOTE]
> Introduces functional expert parallelism for MoE and wires it through model setup, FSDP, optimizer, and training.
> 
> - **EP for MoE layers**: Parallelizes `MoE.experts` via `ExpertParallel` (`apply_ep`), adds EP-aware FSDP sharding on `dp_shard_mod_ep` mesh, sets gradient divide factor via `ParallelDims.fsdp_gradient_divide_factor`, and configures forward/backward prefetch to avoid FSDP interference.
> - **Model setup**: Applies EP before AC/compile/FSDP; refactors FSDP config with `offload_policy` and shared `fsdp_config`; conditional shard of `embed_tokens`/`lm_head`/`norm` preserved via `shard_norm_and_lm_head`.
> - **Optimizer (Muon)**: Switches to `ParallelDims`; adds separate param groups for `mlp.experts` and `mlp.router`, with EP-aware `distributed_mesh_name`, sets `distributed_mesh`, `world_mesh`, and `fsdp_mesh_dim`; updates `MultiLoRAOptimizer` accordingly.
> - **Training**: Uses EP-aware `clip_grad_norm_` and passes `ParallelDims` into optimizer setup in both `rl.train` and `sft.train`.
> - **Config/validation**: Adds guard that `model.ep>1` is allowed only with `impl` in (`custom`, `auto`) for RL and SFT configs.
> - **Changelog & deps**: Documents `model.ep`; pins `dion` to commit `d891eeb` in `pyproject.toml` and `uv.lock`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e5e11ad044e589955d1647f77d97a36be5ac39cd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/model.py [hunk 1]
```diff
@@ -194,6 +214,50 @@ def setup_fsdp(model: nn.Module, config: ModelConfig, parallel_dims: ParallelDim
...
 
+    if not parallel_dims.ep_enabled:
+        return
+
+    transformer_blocks = list(model.model.layers)
+    next_transformer_blocks = transformer_blocks[1:] + [None]
+
+    if model.model.embed_tokens is not None and len(model.model.layers) > 0:
+        if shard_norm_and_lm_head:
+            model.model.embed_tokens.set_modules_to_forward_prefetch([transformer_blocks[0]])
```
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/prime_rl/trainer/model.py:229]: BS

File: src/prime_rl/trainer/rl/train.py [hunk 2]
```diff
@@ -393,11 +398,10 @@ def compute_scaling(orch_config) -> float:
             logger.debug(micro_step_message)
 
         # Optionally, clip the gradients
-        grad_norm_dtensor = torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=config.optim.max_norm)
-        # Convert to CUDA if on CPU (needed for FSDP CPU offloading)
-        if grad_norm_dtensor.device.type == "cpu":
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py:398]: why removing this ?

File: src/prime_rl/trainer/sft/train.py [hunk 3]
```diff
@@ -284,7 +286,9 @@ def train(config: SFTTrainerConfig):
             logger.debug(micro_step_message)
 
         logger.debug(f"Clipping gradients with max norm {config.optim.max_norm}")
-        grad_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=config.optim.max_norm).full_tensor()
+        grad_norm = clip_grad_norm_(
+            model.parameters(), max_norm=config.optim.max_norm, ep_enabled=parallel_dims.ep_enabled
+        ).to(torch.device("cuda"))
```
Comment (samsja) [inline src/prime_rl/trainer/sft/train.py]: not sure if we need the extra .to()

File: src/prime_rl/trainer/sft/train.py [hunk 4]
```diff
@@ -46,6 +46,8 @@
 import torch.distributed as dist
 from liger_kernel.transformers.cross_entropy import LigerCrossEntropyLoss
 
+from torchtitan.distributed.utils import clip_grad_norm_
```
Comment (samsja) [inline src/prime_rl/trainer/sft/train.py:49]: not super fan of using old torchtitan like this, shouldn't bump torchtitan to latest at least ?

File: src/prime_rl/trainer/model.py [hunk 5]
```diff
@@ -194,6 +214,50 @@ def setup_fsdp(model: nn.Module, config: ModelConfig, parallel_dims: ParallelDim
...
+        if prev_transformer_block is not None:
+            if getattr(prev_transformer_block, "moe_enabled", False):
+                transformer_block.set_modules_to_backward_prefetch(
+                    [prev_transformer_block, prev_transformer_block.mlp.experts]
+                )
+            else:
+                transformer_block.set_modules_to_backward_prefetch([prev_transformer_block])
+        elif model.model.embed_tokens is not None:
+            if shard_norm_and_lm_head:
+                transformer_block.set_modules_to_backward_prefetch([model.model.embed_tokens])
```
Comment (samsja) [inline src/prime_rl/trainer/model.py:263]: need comment here not sure to follow what it does

File: src/prime_rl/trainer/models/glm4_moe/modeling_glm4_moe.py [hunk 6]
```diff
@@ -77,6 +77,7 @@ def __init__(self, config: Glm4MoeConfig, layer_idx: int):
 
         if layer_idx >= config.first_k_dense_replace:
             self.mlp = MoE(moe_args, dim=config.hidden_size, hidden_dim=config.moe_intermediate_size)
+            self.moe_enabled = True
```
Comment (samsja) [inline src/prime_rl/trainer/models/glm4_moe/modeling_glm4_moe.py]: hmm how can we remenber that we need to add this all the time to moe layer, wondering if we shouldn't rather look at the class of .mlp to see if its MoE

File: src/prime_rl/trainer/sft/train.py (hunk 4 above)
Comment (S1ro1) [reply src/prime_rl/trainer/sft/train.py:49]: so we can either copy it out or keep it like this, it's a very minimal wrapper around clip_grad_norm_ that decides if it's an EP param or not, doesn't matter to me really, just reply if we should copy it out

File: src/prime_rl/trainer/sft/train.py (hunk 3 above)
Comment (S1ro1) [reply src/prime_rl/trainer/sft/train.py]: it's gonna be a no-op if already on cuda, it can also be on cpu if offloaded

File: src/prime_rl/trainer/rl/train.py (hunk 2 above)
Comment (S1ro1) [reply src/prime_rl/trainer/rl/train.py:398]: it's added back

File: src/prime_rl/trainer/sft/train.py (hunk 4 above)
Comment (samsja) [reply src/prime_rl/trainer/sft/train.py:49]: sure lets keep it like thsi then

**Review by samsja — APPROVED** [review]
lfgtm
